contracts/contracts/strategies/ConvexGeneralizedMetaStrategy.sol

```diff
@@ -0,0 +1,153 @@
+// SPDX-License-Identifier: agpl-3.0
+pragma solidity ^0.8.0;
+
+/**
+ * @title Curve Convex Strategy
+ * @notice Investment strategy for investing stablecoins via Curve 3Pool
+ * @author Origin Protocol Inc
+ */
+import { SafeERC20 } from "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import "@openzeppelin/contracts/utils/Strings.sol";
+
+import { IRewardStaking } from "./IRewardStaking.sol";
+import { IConvexDeposits } from "./IConvexDeposits.sol";
+import { ICurvePool } from "./ICurvePool.sol";
+import { IERC20 } from "./BaseCurveStrategy.sol";
+import { BaseConvexMetaStrategy } from "./BaseConvexMetaStrategy.sol";
+import { StableMath } from "../utils/StableMath.sol";
+
+contract ConvexGeneralizedMetaStrategy is BaseConvexMetaStrategy {
+    using StableMath for uint256;
+    using SafeERC20 for IERC20;
+
+    /* Take 3pool LP and deposit it to metapool. Take the LP from metapool
+     * and deposit them to Convex.
+     */
+    function _lpDepositAll() internal override {
+        IERC20 threePoolLp = IERC20(pTokenAddress);
+        ICurvePool curvePool = ICurvePool(platformAddress);
+
+        uint256 threePoolLpBalance = threePoolLp.balanceOf(address(this));
+        uint256 curve3PoolVirtualPrice = curvePool.get_virtual_price();
+        uint256 threePoolLpDollarValue = threePoolLpBalance.mulTruncate(
+            curve3PoolVirtualPrice
+        );
+
+        uint256[2] memory _amounts = [0, threePoolLpBalance];
+
+        uint256 metapoolVirtualPrice = metapool.get_virtual_price();
+        /**
+         * First convert all the deposited tokens to dollar values,
+         * then divide by virtual price to convert to metapool LP tokens
+         * and apply the max slippage
+         */
+        uint256 minReceived = threePoolLpDollarValue
+            .divPrecisely(metapoolVirtualPrice)
+            .mulTruncate(uint256(1e18) - maxSlippage);
+
+        // slither-disable-next-line unused-return
+        metapool.add_liquidity(_amounts, minReceived);
+
+        uint256 metapoolLp = metapoolLPToken.balanceOf(address(this));
+
+        bool success = IConvexDeposits(cvxDepositorAddress).deposit(
+            cvxDepositorPTokenId,
+            metapoolLp,
+            true // Deposit with staking
+        );
+
+        require(success, "Failed to deposit to Convex");
+    }
+
+    /**
+     * Withdraw the specified amount of tokens from the gauge. And use all the resulting tokens
+     * to remove liquidity from metapool
+     * @param num3CrvTokens Number of Convex LP tokens to remove from gauge
+     */
+    function _lpWithdraw(uint256 num3CrvTokens) internal override {
+        ICurvePool curvePool = ICurvePool(platformAddress);
+
+        uint256 gaugeTokens = IRewardStaking(cvxRewardStakerAddress).balanceOf(
+            address(this)
+        );
+        /**
+         * Convert 3crv tokens to metapoolLP tokens and double it. Doubling is required because aside
```

Comment:
I think these comments aren't correct for this file?

Reply:
ah yes thanks, corrected the comment